File: 202-happy-number/happy-number.java
class Solution {
    public boolean isHappy(int n) {
        int fast = n;
        int slow = n ;
        while(true){
            slow = helper(slow) ;
            fast = helper(fast) ;
            fast = helper(fast) ;

            if(fast == 1){
                return true ;
            }else if(fast == slow){
                return false ;
            }
        }
    }


    private int helper(int n ){
        int sum = 0 ;
        while(n>0){
            int last_digit = n%10 ;

            sum = sum + (last_digit * last_digit) ;
            n = n/10 ;
            

        }
        return sum ;
    }
}

File: 202-happy-number/happy-number.go
func isHappy(n int) bool {
    m := make(map[int] bool)

    for {
        n = helper(n)
        if n == 1{
            return true
        }else if m[n]{
            break ;
        }else{
            m[n] = true
        }
    }

    return false
}

func helper(n int) int{
     sum := 0
     for n>0{
        last_digit := n%10
        sum = sum + (last_digit *last_digit)
        n = n/10
     }
     return sum
}